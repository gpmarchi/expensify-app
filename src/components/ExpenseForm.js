import React from "react";
import moment from "moment";
import "react-dates/initialize";
import { SingleDatePicker } from "react-dates";

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : "",
      amount: props.expense ? (props.expense.amount / 100).toString() : "",
      note: props.expense ? props.expense.note : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: undefined
    };
  }

  handleDescriptionChange = event => {
    const description = event.target.value;
    this.setState(() => ({ description }));
  };

  handleNoteChange = event => {
    const note = event.target.value;
    this.setState(() => ({ note }));
  };

  handleAmountChange = event => {
    const amount = event.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  handleDateChange = createdAt => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };

  handleFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };

  handleSubmit = event => {
    event.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState({ error: "Please provide description and amount" });
    } else {
      this.setState({ error: undefined });
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.handleSubmit}>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.handleDescriptionChange}
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.handleAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.handleDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.handleFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder="Add a note for your expenses (optional)"
            value={this.state.note}
            onChange={this.handleNoteChange}
          />
          <button type="submit">Add Expense</button>
        </form>
      </div>
    );
  }
}
