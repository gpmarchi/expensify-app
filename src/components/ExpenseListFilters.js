import React from "react";
import { connect } from "react-redux";
import "react-dates/initialize";
import { DateRangePicker } from "react-dates";
import {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate
} from "../actions/filters";

export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null
  };

  handleDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  handleFocusChange = calendarFocused => {
    this.setState(() => ({ calendarFocused }));
  };

  handleTextChange = event => {
    this.props.setTextFilter(event.target.value);
  };

  handleSortChange = event => {
    if (event.target.value === "date") {
      this.props.sortByDate();
    } else if (event.target.value === "amount") {
      this.props.sortByAmount();
    }
  };

  render() {
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input
              type="text"
              className="text-input"
              placeholder="Search expenses"
              value={this.props.filters.text}
              onChange={this.handleTextChange}
            />
          </div>
          <div className="input-group__item">
            <select
              className="select"
              value={this.props.filters.sortBy}
              onChange={this.handleSortChange}
            >
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>
          <div className="input-group__item">
            <DateRangePicker
              startDate={this.props.filters.startDate}
              startDateId={
                this.props.filters.startDate
                  ? this.props.filters.startDate.toString()
                  : ""
              }
              endDate={this.props.filters.endDate}
              endDateId={
                this.props.filters.endDate
                  ? this.props.filters.endDate.toString()
                  : ""
              }
              onDatesChange={this.handleDatesChange}
              focusedInput={this.state.calendarFocused}
              onFocusChange={this.handleFocusChange}
              showClearDates={true}
              numberOfMonths={1}
              isOutsideRange={() => false}
            />
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({ filters: state.filters });

const mapDispatchToProps = dispatch => ({
  setTextFilter: text => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: startDate => dispatch(setStartDate(startDate)),
  setEndDate: endDate => dispatch(setEndDate(endDate))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ExpenseListFilters);
