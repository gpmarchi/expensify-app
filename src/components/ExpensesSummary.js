import React from "react";
import { connect } from "react-redux";
import numeral from "numeral";
import selectExpenses from "../selectors/expenses";
import selectExpensesTotal from "../selectors/expenses-total";

export const ExpensesSummary = props => (
  <div>
    <p>
      Viewing {props.expenseCount} expense(s) totalling {props.expensesTotal}
    </p>
  </div>
);

const mapStateToProps = state => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  const visibleExpensesTotal = selectExpensesTotal(visibleExpenses);

  return {
    expenseCount: visibleExpenses.length,
    expensesTotal: numeral(visibleExpensesTotal / 100).format("$0,0.00")
  };
};

export default connect(mapStateToProps)(ExpensesSummary);
