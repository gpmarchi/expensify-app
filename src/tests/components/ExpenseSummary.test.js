import React from "react";
import { shallow } from "enzyme";
import numeral from "numeral";
import { ExpensesSummary } from "../../components/ExpensesSummary";
import expenses from "../fixtures/expenses";
import selectExpensesTotal from "../../selectors/expenses-total";

test("Should render ExpensesSummary with 0 expenses correctly", () => {
  const expenseCount = 0;
  const expensesTotal = numeral(0).format("$0,0.00");
  const wrapper = shallow(
    <ExpensesSummary
      expenseCount={expenseCount}
      expensesTotal={expensesTotal}
    />
  );
  expect(wrapper).toMatchSnapshot();
});

test("Should render ExpensesSummary with 1 expense correctly", () => {
  const expenseCount = 1;
  const expensesTotal = numeral(expenses[0].amount / 100).format("$0,0.00");
  const wrapper = shallow(
    <ExpensesSummary
      expenseCount={expenseCount}
      expensesTotal={expensesTotal}
    />
  );
  expect(wrapper).toMatchSnapshot();
});

test("Should render ExpensesSummary with multiple expenses correctly", () => {
  const expenseCount = expenses.length;
  const expensesTotal = numeral(selectExpensesTotal(expenses) / 100).format(
    "$0,0.00"
  );
  const wrapper = shallow(
    <ExpensesSummary
      expenseCount={expenseCount}
      expensesTotal={expensesTotal}
    />
  );
  expect(wrapper).toMatchSnapshot();
});
