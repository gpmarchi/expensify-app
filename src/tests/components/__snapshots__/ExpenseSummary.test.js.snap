// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpensesSummary with 0 expenses correctly 1`] = `
<div>
  <p>
    Viewing 
    0
     expense(s) totalling 
    $0.00
  </p>
</div>
`;

exports[`Should render ExpensesSummary with 1 expense correctly 1`] = `
<div>
  <p>
    Viewing 
    1
     expense(s) totalling 
    $1.95
  </p>
</div>
`;

exports[`Should render ExpensesSummary with multiple expenses correctly 1`] = `
<div>
  <p>
    Viewing 
    3
     expense(s) totalling 
    $1,141.95
  </p>
</div>
`;
